lib/appsignal/utils/argument_cleaner.ex

```diff
@@ -1,17 +1,124 @@
+defmodule Appsignal.Utils.Literal do
+  defstruct [:value]
+end
+
+defimpl Inspect, for: Appsignal.Utils.Literal do
+  def inspect(%Appsignal.Utils.Literal{value: value}, _opts), do: value
+end
+
 defmodule Appsignal.Utils.ArgumentCleaner do
-  alias Appsignal.Utils.Type
+  @maximum_iterable_count 4
+  @maximum_recursion 2
 
-  def clean(argument)
-      when is_boolean(argument) or is_integer(argument) or is_float(argument) or is_pid(argument) or
-             is_port(argument) or is_reference(argument) do
-    argument
-  end
+  def clean_literal(argument), do: %Appsignal.Utils.Literal{value: clean(argument)}
+  def clean(argument), do: clean(argument, @maximum_recursion)
 
-  def clean(argument) when is_map(argument) do
+  def clean(argument, _) when is_nil(argument), do: inspect(argument)
+  def clean(argument, _) when is_atom(argument), do: inspect(argument)
+  def clean(argument, _) when is_binary(argument), do: "\"...\""
+  def clean(argument, _) when is_bitstring(argument), do: "<<...>>"
+
+  def clean(argument, recurse) when is_map(argument) do
     {struct, map} = Map.pop(argument, :__struct__)
 
-    "%#{Appsignal.Utils.module_name(struct)}{#{Enum.map_join(map, ", ", fn {key, value} -> "#{inspect(key)} => #{inspect(clean(value))}" end)}}"
+    contents = if Enum.empty?(map), do: "", else: do_clean_map(map, recurse, struct)
+
+    if struct do
+      "%#{Appsignal.Utils.module_name(struct)}{#{contents}}"
+    else
+      "%{#{contents}}"
+    end
+  end
+
+  def clean(argument, _) when is_integer(argument), do: "integer()"
+
+  def clean(argument, _) when is_float(argument), do: "float()"
+
+  def clean(argument, _) when is_pid(argument), do: "#PID<...>"
+
+  def clean(argument, _) when is_port(argument), do: "#Port<...>"
+
+  def clean(argument, _) when is_reference(argument), do: "#Reference<...>"
+
+  def clean(argument, _) when is_function(argument) do
+    inspected_function = inspect(argument)
+
+    if String.starts_with?(inspected_function, "#Function<") do
+      # Represent the function as `fn _, _ -> ... end`, as this allows
+      # to express the arity clearly while redacting the rest.
+      arity = :erlang.fun_info(argument)[:arity]
+      placeholders = if arity == 0, do: "", else: Enum.map_join(1..arity, ", ", fn _ -> "_" end)
+      Enum.join(Enum.reject(["fn", placeholders, "-> ... end"], &(&1 == "")), " ")
+    else
+      # As it is not an anonymous function, calling `inspect` on it already
+      # expresses it in the Module.function/arity form, which is fine.
+      inspected_function
+    end
+  end
+
+  def clean(argument, recurse) when is_tuple(argument) do
+    if Enum.empty?(Tuple.to_list(argument)) do
```

Comment:
There are a couple of functions which have an `if`-statement as their function bodies that have a specific case when the value is empty, which I think could be rewritten to use multiple function clauses.

In this example, because we know what an empty tuple looks like, we can add a clause that returns `"{}"` when the argument is `{}` and circumvent the "main" clause. That would make the code more readable, and save a call to `Tuple.to_list/1`.

``` elixir
  def clean({}, _), do: "{}"

  def clean(argument, recurse) when is_tuple(argument) do
    "{#{do_clean_tuple(argument, recurse)}}"
  end
```

Reply:
Oh, that makes a lot of sense -- I didn't realise that I could check for the zero-member case directly with pattern-matching. Will do!

Reply:
Done for tuples and lists. Can't do this for maps, as pattern-matching for maps is non-exhaustive (`%{}` matches any map, not the empty map)